File: resources/patterns/design/queue-based-load-leveling-pattern/queue_consumer.py
"""Queue-Based Load Leveling Pattern - Consumer handler examples."""

import time
import random


def handle_email(task):
    """Simulate sending an email."""
    time.sleep(0.05)
    if random.random() < 0.05:
        raise RuntimeError("SMTP timeout")
    return {"status": "sent", "task_id": task.id}


def handle_report(task):
    """Simulate generating a report."""
    time.sleep(0.1)
    return {"status": "completed", "task_id": task.id, "report_url": f"/reports/{task.id}.pdf"}


def handle_payment(task):
    """Simulate processing a payment."""
    time.sleep(0.08)
    if random.random() < 0.1:
        raise RuntimeError("Payment gateway unavailable")
    return {"status": "processed", "task_id": task.id, "amount": task.payload.get("amount")}


HANDLERS = {
    "send-email": handle_email,
    "generate-report": handle_report,
    "process-payment": handle_payment,
}


def route_task(task):
    """Route a task to the correct handler based on task type."""
    handler = HANDLERS.get(task.type)
    if handler is None:
        raise ValueError(f"Unknown task type: {task.type}")
    return handler(task)

File: resources/patterns/design/queue-based-load-leveling-pattern/queue_consumer.js
// Queue-Based Load Leveling Pattern - Consumer (in-memory, no BullMQ dependency)

const handlers = {
  'send-email': (data) => ({ status: 'sent', to: data.to }),
  'process-payment': (data) => ({ status: 'processed', amount: data.amount }),
  'generate-report': (data) => ({ status: 'completed', type: data.type }),
};

class TaskConsumer {
  constructor(queue, handlerMap, interval = 10) {
    this.queue = queue;
    this.handlers = handlerMap || handlers;
    this.interval = interval;
    this.running = false;
    this.processed = 0;
    this.failed = 0;
    this.timer = null;
  }

  start(concurrency = 1) {
    this.running = true;
    this._loop();
  }

  _loop() {
    if (!this.running) return;
    const task = this.queue.dequeue();
    if (task) {
      try {
        const handler = this.handlers[task.type];
        if (!handler) throw new Error(`Unknown task type: ${task.type}`);
        handler(task.payload);
        this.processed++;
      } catch (err) {
        this.failed++;
        this.queue.fail(task, err.message);
      }
    }
    this.timer = setTimeout(() => this._loop(), this.interval);
  }

  stop() {
    this.running = false;
    if (this.timer) clearTimeout(this.timer);
  }

  getStats() {
    return { processed: this.processed, failed: this.failed };
  }
}

function routeTask(task) {
  const handler = handlers[task.type];
  if (!handler) throw new Error(`Unknown task type: ${task.type}`);
  return handler(task.payload);
}

module.exports = { TaskConsumer, routeTask, handlers };
